### ISSUE TITLE
calling ctx.expand_location() from an aspect causes outputs of rule to be required

### ISSUE THREAD

**Issue by ob** [bazelbuild/bazel#7083]
If an aspect rule calls `ctx.expand_location()` for an attribute of the aspect's `ctx.rule`, the `ctx.rule`'s outputs are added to the _required outputs_ list. This makes the build fail since the actions to create the outputs are not necessarily created.

Here is a shell script that reproduces the issue:

```
#!/bin/sh

set -eo pipefail

workdir=$(mktemp -d)
cd $workdir
touch WORKSPACE
mkdir simple
mkdir aspects
cat >simple/hello.m <<'EOF'
#include <stdio.h>

void
say_hello()
{
  printf("Hello\n");
}
EOF

cat >simple/BUILD <<'EOF'
load(":headermap.bzl", "headermap")
headermap(
    name = "headermap",
)

objc_library(
    name = "simple",
    srcs = glob(["*.m"]),
    hdrs = glob(["*.h"]),
    copts = ["-I$(location headermap)"],
    deps = ["headermap"],
    visibility = ["//visibility:public"],
)
EOF

cat >simple/headermap.bzl <<'EOF'
def _headermap_impl(ctx):
    ctx.actions.write(
        content="",
        output=ctx.outputs.headermap,
    )

    # Allow objc rules to depend on this
    objc_provider = apple_common.new_objc_provider(
        header = depset([ctx.outputs.headermap]),
    )
    return struct(
        files=depset([ctx.outputs.headermap]),
        providers = [objc_provider],
        objc = objc_provider,
        headers = depset([ctx.outputs.headermap]),
    )

headermap = rule(
    implementation = _headermap_impl,
    output_to_genfiles = True,
    attrs = { },
    outputs = {"headermap": "%{name}.hmap"}
)
EOF

touch aspects/BUILD
cat >aspects/print.bzl <<'EOF'
def _print_expanded_copts_impl(target, ctx):
    # Make sure the rule has a srcs attribute.
    if hasattr(ctx.rule.attr, 'copts'):
        # Iterate through the files that make up the sources and
        # print their paths.
        for opt in ctx.rule.attr.copts:
            print("{} -> {}".format(opt, ctx.expand_location(opt, [])))
    return []

print_expanded_copts = aspect(
    implementation = _print_expanded_copts_impl,
    attr_aspects = ['deps'],
)
EOF
bazel clean --expunge ; bazel build simple:simple --aspects aspects/print.bzl%print_expanded_copts
```

Saving this shell script in `repro.sh` and running it produces the following output:

```
$ sh repro.sh 
Starting local Bazel server and connecting to it...
INFO: Invocation ID: bc62eba1-2daa-4c45-af6f-e776a5a4e9db
INFO: Starting clean.
Starting local Bazel server and connecting to it...
INFO: Invocation ID: cca655ee-3334-41d3-b289-2ceb80a48ff3
DEBUG: /private/var/folders/b2/5s3n8vxx2wx22cm0x66dby1c000ncy/T/tmp.Txwhwu1W/aspects/print.bzl:7:13: -I$(location headermap) -> -Ibazel-out/apl-ios_x86_64-fastbuild/genfiles/simple/headermap.hmap
ERROR: /private/var/folders/b2/5s3n8vxx2wx22cm0x66dby1c000ncy/T/tmp.Txwhwu1W/simple/BUILD:6:1: in //aspects:print.bzl%print_expanded_copts aspect on objc_library rule //simple:simple: 


The following files have no generating action:
simple/simple_fully_linked.a
ERROR: Analysis of aspect '//aspects:print.bzl%print_expanded_copts of //simple:simple' failed; build aborted: Analysis of target '//simple:simple' failed; build aborted
INFO: Elapsed time: 9.467s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (17 packages loaded, 329 targets configured)
```

The failure is due to the outputs of the `simple` rule being added to the list of required outputs, but note that `simple` is not required to expand the location, it's `headermap` that is required.

Tested with Bazel 0.21.0 and master as of 08cba48d21e66af1396742ad9ee6c94c3be19132 on macOS 10.14.2.

**Comment by ob** [comment]
So looking a bit more into this bug, I think the problem is around [here](https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/analysis/LocationExpander.java#L367), where the code to expand locations creates the artifacts via a call to `CreateOutputArtifact()`. This has a side effect of adding the artifact to the list of tracked artifacts in the CachingAnalysisEnvironment, [here](https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/analysis/CachingAnalysisEnvironment.java#L254). 

So if you look at the flow, the action is registered in on Analysis Environment, here _d342bd6_:

```
DEBUG: Registering action 'action 'Linking simple/simple_fully_linked.a' in 'com.google.devtools.build.lib.analysis.CachingAnalysisEnvironment@d342bd6'
```

And then the artifact is looked up in a different Analysis Environment (_14773d5d_):

```
DEBUG: Aspect Created: 'com.google.devtools.build.lib.analysis.ConfiguredAspect@1b68d16c' in rule context RuleContext(//simple:headermap, e28ed302fddd6943df6571ce00b5fe73)'
DEBUG: Adding artifact bazel-out/apl-darwin_x86_64-fastbuild/bin/simple/simple_fully_linked.a to 'com.google.devtools.build.lib.analysis.CachingAnalysisEnvironment@14773d5d' AE
DEBUG: Aspect Created: 'com.google.devtools.build.lib.analysis.ConfiguredAspect@8d4f0d0' in rule context RuleContext(//simple:simple, e28ed302fddd6943df6571ce00b5fe73)'
DEBUG: Could not find bazel-out/apl-darwin_x86_64-fastbuild/bin/simple/simple_fully_linked.a in 'com.google.devtools.build.lib.analysis.CachingAnalysisEnvironment@14773d5d' AE
```

My question is, is there a way in that call to `trackArtifactAndOrigin()` to know you're actually in the aspect graph and not in the action graph? Could I wire in something from the `ruleContext` down to the `CachingAnalysisEnvironment` to avoid this issue?

**Comment by laurentlb** [comment]
cc @c-parsons

**Comment by ob** [comment]
Fixed by commit 5adde8ecc3b45b2f39241d2271019f15bc0c7240 in #7282
